framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,372 @@
+     * @param checker a base type checker or its subclass
+     */
+    public AnnotatedTypeLoader(BaseTypeChecker checker) {
+        this.checker = checker;
+        processingEnv = checker.getProcessingEnvironment();
+        packageName = checker.getClass().getPackage().getName().replace(SLASH, DOT) + QUAL_PACKAGE_SUFFIX;
+        ProcessPackageNameAndLoadAnnos();
+    }
+
+    /**
+     * Constructor for loading type qualifiers defined for an annotated type factory
+     * @param factory an annotated type factory or its subclass
+     */
+    public AnnotatedTypeLoader(AnnotatedTypeFactory factory) {
+        processingEnv = factory.getProcessingEnv();
+        packageName = factory.getClass().getPackage().getName().replace(SLASH, DOT) + QUAL_PACKAGE_SUFFIX;
+        ProcessPackageNameAndLoadAnnos();
+    }
+
+    // sees if the package name of the qual directory is a valid resource and
+    // retrieve its resource URL (either a jar or regular file directory), then
+    // proceeds to reflectively load annotations if it is valid. Returns an
+    // empty set if it isn't a valid URL
+    private void ProcessPackageNameAndLoadAnnos() {
+        //resourceURL = Thread.currentThread().getContextClassLoader().getResource(packageName.replace(DOT, SLASH));
+        resourceURL = this.getClass().getClassLoader().getResource(packageName.replace(DOT, SLASH));
+
+        // in checkers, there will be a resource URL for the qual directory. But
+        // when called in the framework (eg GeneralAnnotatedTypeFactory), there
+        // won't be a resourceURL since there isn't a qual directory.
+        if (resourceURL != null) {
+            // if there's a URL, load from it
+            loadedAnnotations = loadCheckerFrameworkAnnotations();
+        } else {
+            // otherwise create an empty set
+            loadedAnnotations = new HashMap<Class<? extends Annotation>, AnnotationMirror>();
+        }
+    }
+
+    /**
+     * Every subclass of AnnotatedTypeLoader must implement how it will convert a class into an annotation mirror
+     * @param annoClass an annotation's class
+     * @return AnnotationMirror of the annotation, or null if this annotation isn't required
+     */
+    protected AnnotationMirror createAnnotationMirrorFromClass(Class<? extends Annotation> annoClass) {
+        // build the annotation mirror
+        AnnotationBuilder builder = new AnnotationBuilder(processingEnv, annoClass);
+        AnnotationMirror annoMirroResult = builder.build();
+        return annoMirroResult;
+    }
+
+    /**
+     * gets the qualifier set from the loaded annotations, with option to skip {@link PolyAll}
+     * @param excludePolyAll setting this to true would skip adding {@link PolyAll} to the list of qualifiers
+     * @return an unmodifiable set of the qualifiers
+     */
+    public final Set<Class<? extends Annotation>> getAnnotatedTypeQualifierSet(final boolean excludePolyAll) {
```

Comment:
Don't add PolyAll here.  Do that in AnnotatedTypeFactory.loadTypeQualifiersFromQualDir